File: examples/elixir-phoenix/.harmont/pipeline.ts
import { pipeline, push, pullRequest, type PipelineDefinition } from "harmont";
import { elixir } from "harmont/toolchains";

const project = elixir({
  path: ".",
  elixirVersion: "1.18.3",
  otpVersion: "27.3.3",
});

const pipelines: PipelineDefinition[] = [
  {
    slug: "ci",
    triggers: [push({ branch: "main" }), pullRequest()],
    pipeline: pipeline(
      [
        project.compile(),
        project.test({ cover: true }),
        project.format(),
        project.credo(),
        project.dialyzer(),
        project.sobelow(),
        project.depsAudit(),
        project.hexAudit(),
      ],
      {
        env: {
          CI: "true",
          MIX_ENV: "test",
        },
        defaultImage: "ubuntu:24.04",
      },
    ),
  },
  {
    slug: "deploy",
    triggers: [push({ branch: "main" })],
    pipeline: pipeline(
      [project.compile(), project.mix("assets.deploy"), project.release()],
      { env: { MIX_ENV: "prod" }, defaultImage: "ubuntu:24.04" },
    ),
  },
];

export default pipelines;

File: examples/elixir-phoenix/.hm/pipeline.py
"""Elixir Phoenix example pipeline."""
from __future__ import annotations

import harmont as hm


@hm.target()
def project():
    return hm.elixir(path=".", elixir_version="1.18.3", otp_version="27.3.3")


@hm.pipeline(
    "ci",
    env={"CI": "true", "MIX_ENV": "test"},
    triggers=[hm.push(branch="main"), hm.pr()],
)
def ci(project: hm.Target) -> tuple[hm.Step, ...]:
    return (
        project.compile(),
        project.test(cover=True),
        project.format(),
        project.credo(),
        project.dialyzer(),
        project.sobelow(),
        project.deps_audit(),
        project.hex_audit(),
    )


@hm.pipeline(
    "deploy",
    env={"MIX_ENV": "prod"},
    triggers=[hm.push(branch="main")],
)
def deploy(project: hm.Target) -> tuple[hm.Step, ...]:
    return (
        project.compile(),
        project.mix("assets.deploy"),
        project.release(),
    )
